src/openMVG/sfm/sfm_data_filters.cpp

```diff
@@ -193,89 +151,86 @@ IndexT RemoveOutliers_AngleError
 
-    const int num_obs = obs.obs.size();
-    rays.resize(num_obs);
+    objJPts.resize( obs.size() );
 
-    // Here we examine the track and its observations.
-    // If we find that the angle between any pair of observations is acceptable, we immediately advance to the next track.
-    // Otherwise, we will refine the test, potentially looking at all pairs of observations.  If we eventually find
-    // that none are acceptable, we remove the track.
-    for (size_t i = 0; i != num_obs; ++i)
-    {
-      // Build rays as needed
-      if (!rays[i].first)
+    // Remember, we are reusing the store; however, we need to
+    // reset the cached entries for the next iteration too.
+    for (auto & i : objJPts)
     {
-        const auto& obsIt = obs.obs[i];
-        const View* view = views.find(obsIt.first)->second.get();
-      const auto& pi = poseInfo.find(view)->second;
-
-        rays[i] = {
-          true, pi.first * pi.second->oneBearing(pi.second->get_ud_pixel(obsIt.second.x)).normalized()
-        };
+      std::get<0>(i) = false;
     }
-
-      for (size_t j = i+1; j != num_obs; ++j)
-      {
-        // Build as needed
-        if (!rays[j].first)
+    int idxI = 0;
+    // WIP Could rewrite the loop to avoid the last obs calculation.
+    for (Observations::const_iterator itObs1 = obs.begin();
+      itObs1 != obs.end(); ++itObs1)
     {
-          const auto& obsIt = obs.obs[j];
-          const View* view = views.find(obsIt.first)->second.get();
-          const auto& pi = poseInfo.find(view)->second;
+      const View * view1 = views.at(itObs1->first).get();
+      // Getting the post incurs a modest penalty (20%)
+      const geometry::Pose3 pose1 = sfm_data.GetPoseOrDie(view1);
+      const cameras::IntrinsicBase * intrinsic1 = intrinsics.at(view1->id_intrinsic).get();
+      const Vec2 objIPt = intrinsic1->get_ud_pixel(itObs1->second.x);
+
+      Observations::const_iterator itObs2 = itObs1;
+      ++itObs2;
+      int idxJ = idxI+1;
+      for (; itObs2 != obs.end(); ++itObs2)
+      {
+        auto & cachedObjJInfo = objJPts[ idxJ ];
+        if (!std::get<0>( cachedObjJInfo))
+        {
+          const View * view2 = views.at(itObs2->first).get();
+          std::get<1>(cachedObjJInfo) = sfm_data.GetPoseOrDie(view2);
+          const cameras::IntrinsicBase * intrinsic2 = intrinsics.at(view2->id_intrinsic).get();
+          std::get<2>(cachedObjJInfo) = intrinsic2;
+          std::get<3>(cachedObjJInfo) = intrinsic2->get_ud_pixel(itObs2->second.x);
+          std::get<0>(cachedObjJInfo) = true;
+        }
+
+        const double angle = AngleBetweenRay(
+          pose1, intrinsic1,
+          std::get<1>(cachedObjJInfo), std::get<2>(cachedObjJInfo),
+          objIPt, std::get<3>(cachedObjJInfo)
+        );
 
-          rays[j] ={
-            true, pi.first * pi.second->oneBearing(pi.second->get_ud_pixel(obsIt.second.x)).normalized()
-          };
+        if (angle >= dMinAcceptedAngle)
+        {
```

Comment:
Is the goto faster than just setting a flag (like bEarlyOut) and breaking out of the loop?

Reply:
As long as there definitely won't be any exceptions thrown that the goto wouldn't handle properly then it's fine.